### ISSUE TITLE
Pynndescent in t-SNE with k=15 seems a lot slower than in UMAP

### ISSUE THREAD

**Issue by dkobak** [pavlin-policar/openTSNE#146]
I've just noticed that UMAP 0.4.6 with pynndescent 0.5 can be quite a bit faster to construct the kNN graph compared to openTSNE with uniform kernel and k=15. This does not make sense to me. I added verbose=1 to the pynndescent call within openTSNE and here is what I see:

UMAP - 34 sec:

```
UMAP(angular_rp_forest=True, metric='cosine', verbose=1)
Construct fuzzy simplicial set
Mon Oct 19 13:54:22 2020 Finding Nearest Neighbors
Mon Oct 19 13:54:22 2020 Building RP forest with 21 trees
Mon Oct 19 13:54:32 2020 metric NN descent for 17 iterations
	 1  /  17
	 2  /  17
	 3  /  17
	 4  /  17
	 5  /  17
	 6  /  17
	Stopping threshold met -- exiting after 6 iterations
Mon Oct 19 13:54:58 2020 Finished Nearest Neighbor Search
```

tSNE -- 61 sec

```
===> Finding 15 nearest neighbors using NN descent approximate search using cosine distance...
Mon Oct 19 14:03:15 2020 Building RP forest with 21 trees
Mon Oct 19 14:03:33 2020 metric NN descent for 17 iterations
	 1  /  17
	 2  /  17
	 3  /  17
	 4  /  17
	 5  /  17
	 6  /  17
	Stopping threshold met -- exiting after 6 iterations
   --> Time elapsed: 61.53 seconds
```

I ran both snippets twice to make sure all numba code was pre-compiled.

I don't know why this happens. The `pynndescent` call in UMAP looks identical to the one in t-SNE. https://github.com/lmcinnes/umap/blob/master/umap/umap_.py#L290

**Comment by pavlin-policar** [comment]
I see the function calls differ slightly, but it's not obvious from the code. UMAP calls pynndescent with `low_memory=False`, while we here don't set it at all, so we use the default, which is `low_memory=True`. Could this be the culprit?

**Comment by dkobak** [comment]
Good catch, but it does not look like it's the culprit.

```
n_trees = 5 + int(round((data.shape[0]) ** 0.5 / 20))
n_iters = max(5, int(round(np.log2(data.shape[0]))))

%time pynndescent.NNDescent(data, n_neighbors=16, metric='cosine', verbose=1, max_candidates=60, n_trees=n_trees, n_iters=n_iters, low_memory=False)

%time pynndescent.NNDescent(data, n_neighbors=16, metric='cosine', verbose=1, max_candidates=60, n_trees=n_trees, n_iters=n_iters, low_memory=True)

%time affinity.Uniform(data, k_neighbors=15, metric='cosine', verbose=1)
```

I get ~35 sec with both `pynndescent` calls but 60 sec with `affinity.Uniform` call, even though all time is spent on `pynndescent.NNDescent` call from within.

Hmm it's weird. Maybe it's something strange with my machine here.

**Comment by dkobak** [comment]
The plot thickens!

I looked at the system monitor and noticed that when I run `pynndescent.NNDescent` as written above, I have all cores active, but from within openTSNE only 1 core is active. Not a surprise because we call `pynndescent` with `n_jobs=1` when the input is sparse (because it did not use to work otherwise).

The default is `n_jobs=None` which I thought is the same as `n_jobs=1` but apparently it is actually the same as `n_jobs=-1`.

Anyway - the solution for this is to remove the ugly workaround at https://github.com/pavlin-policar/openTSNE/blob/master/openTSNE/nearest_neighbors.py#L384 and use `self.n_jobs`. I tried, and it works beautifully!
